truffle/src/main/java/org/jruby/truffle/nodes/ext/BigDecimalNodes.java

```diff
@@ -178,175 +217,327 @@ public static boolean isNan(RubyBasicObject value) {
             return getBigDecimalType(value) == Type.NAN;
         }
 
+        private void setupCreateBigDecimal() {
+            if (createBigDecimal == null)
+                createBigDecimal = insert(CreateBigDecimalNodeFactory.create(getContext(), getSourceSection(), null, null));
+        }
+
+        protected RubyBasicObject createBigDecimal(VirtualFrame frame, Object value) {
+            setupCreateBigDecimal();
+            return createBigDecimal.executeCreate(frame, value);
+        }
+
+        protected RubyBasicObject createBigDecimal(VirtualFrame frame, Object value, RubyBasicObject self) {
+            setupCreateBigDecimal();
+            return createBigDecimal.executeCreate(frame, value, self);
+        }
+
+        private void setupLimitCall() {
+            if (limitCall == null)
+                limitCall = insert(DispatchHeadNodeFactory.createMethodCall(getContext()));
```

Comment:
Shouldn't there be `CompilerDirectives.transferToInterpreterAndInvalidate();`? I had it there but it seems that all the graph manipulation method have this call on the first line so I've thought I'll remove the duplication.